types/wlr_cursor.c

```diff
@@ -0,0 +1,206 @@
+#include <wlr/types/wlr_cursor.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <wlr/util/log.h>
+#include <wayland-server.h>
+#include <wlr/types/wlr_output.h>
+#include <wlr/types/wlr_output_layout.h>
+#include <wlr/types/wlr_input_device.h>
+
+struct wlr_cursor_device {
+	struct wlr_cursor *cursor;
+	struct wlr_input_device *device;
+	struct wl_list link;
+
+	struct wl_listener motion;
+	struct wl_listener motion_absolute;
+	struct wl_listener button;
+	struct wl_listener axis;
+};
+
+struct wlr_cursor_state {
+	struct wl_list devices;
+	struct wlr_output_layout *layout;
+	struct wlr_xcursor *xcursor;
+};
+
+struct wlr_cursor *wlr_cursor_init() {
+	struct wlr_cursor *cur = calloc(1, sizeof(struct wlr_cursor));
+	if (!cur) {
+		wlr_log(L_ERROR, "Failed to allocate wlr_cursor");
+		return NULL;
+	}
+
+	cur->state = calloc(1, sizeof(struct wlr_cursor_state));
+	if (!cur->state) {
+		wlr_log(L_ERROR, "Failed to allocate wlr_cursor_state");
+		return NULL;
+	}
+
+	wl_list_init(&cur->state->devices);
+
+	wl_signal_init(&cur->events.motion);
+	wl_signal_init(&cur->events.motion_absolute);
+	wl_signal_init(&cur->events.button);
+	wl_signal_init(&cur->events.axis);
+
+	cur->x = 100;
+	cur->y = 100;
+
+	return cur;
+}
+
+void wlr_cursor_destroy(struct wlr_cursor *cur) {
+	struct wlr_cursor_device *device;
+	wl_list_for_each(device, &cur->state->devices, link) {
+		wl_list_remove(&device->link);
+		free(device);
+	}
+
+	free(cur);
+}
+
+void wlr_cursor_set_xcursor(struct wlr_cursor *cur, struct wlr_xcursor *xcur) {
+	cur->state->xcursor = xcur;
+}
+
+void wlr_cursor_warp(struct wlr_cursor *cur, double x, double y) {
+}
+
+void wlr_cursor_move(struct wlr_cursor *cur, double delta_x, double delta_y) {
+	//struct wlr_output *current_output;
+	//current_output = wlr_output_layout_output_at(cur->state->layout, cur->x, cur->y);
+
+	// TODO handle no layout
+	assert(cur->state->layout);
+	// TODO handle layout boundaries
+	double new_x = cur->x + delta_x;
+	double new_y = cur->y + delta_y;
+	int hotspot_x = 0;
+	int hotspot_y = 0;
+
+	if (cur->state->xcursor && cur->state->xcursor->image_count > 0) {
+		struct wlr_xcursor_image *image = cur->state->xcursor->images[0];
+		hotspot_x = image->hotspot_x;
+		hotspot_y = image->hotspot_y;
+	}
+
+	struct wlr_output *output;
+	output = wlr_output_layout_output_at(cur->state->layout, new_x, new_y);
+
+	if (output) {
+		int output_x = new_x;
+		int output_y = new_y;
+
+		// TODO fix double to int rounding issues
+		wlr_output_layout_output_coords(cur->state->layout, output, &output_x, &output_y);
+		wlr_output_move_cursor(output, output_x - hotspot_x, output_y - hotspot_y);
```

Comment:
this call segfaulted for me after unplugging an output while the cursor was still in there. Took a couple of unplugs but got this, haven't investigated at all:
```==29197==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000000 (pc 0x7ff9ec59c2c8 bp 0x7ffd7e20e560 sp 0x7ffd7e20e540 T0)
==29197==The signal is caused by a READ memory access.
==29197==Hint: address points to the zero page.
    #0 0x7ff9ec59c2c7 in legacy_crtc_move_cursor ../backend/drm/drm-legacy.c:51
    #1 0x7ff9ec597f09 in wlr_drm_output_move_cursor ../backend/drm/drm.c:680
    #2 0x7ff9ec5bc901 in wlr_output_move_cursor ../types/wlr_output.c:171
    #3 0x7ff9ec5befdd in wlr_cursor_move ../types/wlr_cursor.c:97
    #4 0x402c17 in handle_cursor_motion ../examples/pointer.c:104
    #5 0x7ff9ec5be67c in wl_signal_emit /usr/include/wayland-server-core.h:388
    #6 0x7ff9ec5bf152 in handle_pointer_motion ../types/wlr_cursor.c:107
    #7 0x7ff9ec5a1346 in wl_signal_emit /usr/include/wayland-server-core.h:388
    #8 0x7ff9ec5a16de in handle_pointer_motion ../backend/libinput/pointer.c:37
    #9 0x7ff9ec5a0aa3 in wlr_libinput_event ../backend/libinput/events.c:202
    #10 0x7ff9ec59f0f8 in wlr_libinput_readable ../backend/libinput/backend.c:35
    #11 0x7ff9ec360c51 in wl_event_loop_dispatch (/lib64/libwayland-server.so.0+0x9c51)
    #12 0x7ff9ec35f499 in wl_display_run (/lib64/libwayland-server.so.0+0x8499)
    #13 0x4040fa in main ../examples/pointer.c:203
    #14 0x7ff9ebb584d9 in __libc_start_main (/lib64/libc.so.6+0x204d9)
    #15 0x402189 in _start (/home/shared/wayland/wlroots/build/examples/pointer+0x402189)
```